src/plugins/ucc_discovery_plugin.cpp

```diff
@@ -0,0 +1,345 @@
+            return;
+          }
+        } else {
+          Fail("The given segment type is not supported for the discovery of UCCs.");
+        }
+      }
+
+      // We save UCC constraints directly inside the table so they can be forwarded to nodes in a query plan.
+      message << " [confirmed in " << format_duration(candidate_time.lap()) << "]";
+      Hyrise::get().log_manager.add_message("UccDiscoveryPlugin", message.str(), LogLevel::Info);
+      table->add_soft_key_constraint(TableKeyConstraint(std::unordered_set(std::initializer_list<ColumnID>{col_id}),
+                                                        KeyConstraintType::UNIQUE, current_commit));
+    });
+  }
+  Hyrise::get().log_manager.add_message("UccDiscoveryPlugin", "Clearing LQP and PQP cache...", LogLevel::Debug);
+
+  Hyrise::get().default_lqp_cache->clear();
+  Hyrise::get().default_pqp_cache->clear();
+}
+
+void UccDiscoveryPlugin::stop() {}
+
+std::vector<std::pair<PluginFunctionName, PluginFunctionPointer>>
+UccDiscoveryPlugin::provided_user_executable_functions() const {
+  return {{"DiscoverUCCs", [&]() { discover_uccs(); }}};
+}
+
+std::shared_ptr<std::vector<UCCCandidate>> UccDiscoveryPlugin::generate_valid_candidates(
+    std::shared_ptr<AbstractLQPNode> root_node, std::shared_ptr<LQPColumnExpression> column_candidate) const {
+  auto candidates = std::make_shared<std::vector<UCCCandidate>>();
+
+  if (!root_node) {
+    // input node may already be nullptr in case we try to get right input of node with only one input
+    return nullptr;
+  } else {
+    visit_lqp(root_node, [&](auto& node) {
+      if (node->type != LQPNodeType::Predicate) {
+        return LQPVisitation::VisitInputs;
+      }
+
+      // when looking at predicate node, check whether the searched column is filtered in this predicate
+      // -> if so, it is a valid UCC candidate; if not, still continue search
+      const auto casted_node = std::static_pointer_cast<PredicateNode>(node);
+
+      // first, ensure that we look at a binary predicate expression checking for equality (e.g., A==B)
+      const auto predicate = std::dynamic_pointer_cast<BinaryPredicateExpression>(casted_node->predicate());
+      if ((!predicate) || (predicate->predicate_condition != PredicateCondition::Equals)) {
+        return LQPVisitation::VisitInputs;
+      }
+
+      // get the column expression, should be left, but also check the right operand if the left one is not column
+      auto column_expression = std::dynamic_pointer_cast<LQPColumnExpression>(predicate->left_operand());
+      auto value_expression = std::dynamic_pointer_cast<ValueExpression>(predicate->right_operand());
+      if (!column_expression) {
+        column_expression = std::dynamic_pointer_cast<LQPColumnExpression>(predicate->right_operand());
+        value_expression = std::dynamic_pointer_cast<ValueExpression>(predicate->left_operand());
+      }
```

Comment:
So, the code below works, however, despite using the ``visit_expression`` function it is longer, requires more casts and is, in my opinion, not much easier to understand.

The issue is, that we already casted the ``predicate`` to a ``BinaryPredicateExpression``, as we need to know whether the predicate is using the == condition. However, the ``visit_expression`` function requires a shared pointer to an ``AbstractExpression``, so it throws compilation errors when simply plugging in the previously casted ``predicate``.

The function itself is then fine, but as I said - I don't really see the advantage over the swapping which side of the operand to cast.

```
std::shared_ptr<LQPColumnExpression> column_expression = nullptr;
std::shared_ptr<ValueExpression> value_expression = nullptr;

const auto abstreact_expression_predicate = std::dynamic_pointer_cast<AbstractExpression>(predicate);
visit_expression(abstreact_expression_predicate, [&](auto& sub_expression) {
  if (sub_expression->type == ExpressionType::LQPColumn) {
    column_expression = std::static_pointer_cast<LQPColumnExpression>(sub_expression);
  } else if (sub_expression->type == ExpressionType::Value) {
    value_expression = std::static_pointer_cast<ValueExpression>(sub_expression);
  }
  return ExpressionVisitation::VisitArguments;
});
```

As for the loop - I am not quite sure whether that would improve the code either, since we already know that there will be a left and a right operand and that they have to be one column and one value expression.